<a href="https://colab.research.google.com/github/Bibek04-bit/Flood-Forecasting/blob/main/Data_set_preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving final_flood_features (1).csv to final_flood_features (1).csv


In [4]:
import pandas as pd
df = pd.read_csv("final_flood_features (1).csv")

In [6]:
df.head()

Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0  2009-01-01            470.0               3.36  0.768700             NaN   
1  2009-01-02            460.0               3.34  0.766881             NaN   
2  2009-01-03            456.0               3.33  0.765062             NaN   
3  2009-01-04            448.0               3.31  0.763244             NaN   
4  2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  
0             NaN              0.0      20.222000   63.263360  
1             NaN              0.0      18.118000   54.073805  
2             NaN              0.0      22.784667   65.749364  
3             NaN              0.0      22.027667   68.656034  
4             NaN              0.0      18.637667   58.216001

In [22]:
output_filename = 'final_engineered_flood_data.csv'
df.to_csv(output_filename, index=False)
print(f"DataFrame saved to '{output_filename}'")

DataFrame saved to 'final_engineered_flood_data.csv'


In [23]:
from google.colab import files
files.download(output_filename)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [7]:
df['Date'] = pd.to_datetime(df['Date'])
df = df.sort_values(by='Date').reset_index(drop=True)
print("DataFrame after converting 'Date' to datetime and sorting:")
df.head()

DataFrame after converting 'Date' to datetime and sorting:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  
0             NaN              0.0      20.222000   63.263360  
1             NaN              0.0      18.118000   54.073805  
2             NaN              0.0      22.784667   65.749364  
3             NaN              0.0      22.027667   68.656034  
4             NaN              0.0      18.637667   58.216001

In [8]:
df['flow_lag_1d'] = df['Daily flow Rate'].shift(1)
df['flow_lag_3d'] = df['Daily flow Rate'].shift(3)
print("DataFrame with new 'flow_lag_1d' and 'flow_lag_3d' features:")
df.head()

DataFrame with new 'flow_lag_1d' and 'flow_lag_3d' features:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  
0          NaN  
1          NaN  
2          NaN  
3        470.0  
4        460.0

In [9]:
df['flow_change_rate'] = df['Daily flow Rate'] - df['flow_lag_1d']
print("DataFrame with new 'flow_change_rate' feature:")
df.head()

DataFrame with new 'flow_change_rate' feature:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  
0          NaN               NaN  
1          NaN             -10.0  
2          NaN              -4.0  
3        470.0              -8.0  
4        460.0              -3.0

In [10]:
decay_factor = 0.85
df['API'] = 0.0

# Initialize the first API value
df.loc[0, 'API'] = df.loc[0, '24h accumulated']

# Calculate API for subsequent days
for i in range(1, len(df)):
    df.loc[i, 'API'] = (df.loc[i-1, 'API'] * decay_factor) + df.loc[i, '24h accumulated']

print("DataFrame with new 'API' feature:")
df.head()

DataFrame with new 'API' feature:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  
0          NaN               NaN  0.0  
1          NaN             -10.0  0.0  
2          NaN              -4.0  0.0  
3        470.0              -8.0  0.0  
4        460.0              -3.0  0.0

In [11]:
df['24h_accumulated_roll_3d'] = df['24h accumulated'].rolling(window=3).sum()
df['24h_accumulated_roll_5d'] = df['24h accumulated'].rolling(window=5).sum()
df['24h_accumulated_roll_7d'] = df['24h accumulated'].rolling(window=7).sum()
print("DataFrame with new rolling sum features for '24h accumulated':")
df.head(10)

DataFrame with new rolling sum features for '24h accumulated':


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   
5 2009-01-06            449.0               3.31  0.759606             NaN   
6 2009-01-07            453.0               3.32  0.757787             NaN   
7 2009-01-08            449.0               3.31  0.755969             NaN   
8 2009-01-09            441.0               3.29  0.754150             NaN   
9 2009-01-10            433.0               3.27  0.752331             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   
5             NaN              0.0      17.945667   56.251036        445.0   
6             NaN              0.0      18.687000   58.913774        449.0   
7             NaN              0.0      17.945667   56.251036        453.0   
8             NaN              0.0      17.719000   53.985001        449.0   
9             NaN              0.0      19.164333   60.726332        441.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   
5        456.0               4.0  0.0                      0.0   
6        448.0               4.0  0.0                      0.0   
7        445.0              -4.0  0.0                      0.0   
8        449.0              -8.0  0.0                      0.0   
9        453.0              -8.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  
0                      NaN                      NaN  
1                      NaN                      NaN  
2                      NaN                      NaN  
3                      NaN                      NaN  
4                      0.0                      NaN  
5                      0.0                      NaN  
6                      0.0                      0.0  
7                      0.0                      0.0  
8                      0.0                      0.0  
9                      0.0                      0.0

In [12]:
df['saturation_index'] = (df['sm_top_0_10cm'] + df['sm_10_40cm']) / 2
print("DataFrame with new 'saturation_index' feature:")
df.head()

DataFrame with new 'saturation_index' feature:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  
0                      NaN                      NaN         41.742680  
1                      NaN                      NaN         36.095903  
2                      NaN                      NaN         44.267015  
3                      NaN                      NaN         45.341850  
4                      0.0                      NaN         38.426834

In [13]:
print("DataFrame with all engineered features:")
df.head(10)

DataFrame with all engineered features:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   
5 2009-01-06            449.0               3.31  0.759606             NaN   
6 2009-01-07            453.0               3.32  0.757787             NaN   
7 2009-01-08            449.0               3.31  0.755969             NaN   
8 2009-01-09            441.0               3.29  0.754150             NaN   
9 2009-01-10            433.0               3.27  0.752331             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   
5             NaN              0.0      17.945667   56.251036        445.0   
6             NaN              0.0      18.687000   58.913774        449.0   
7             NaN              0.0      17.945667   56.251036        453.0   
8             NaN              0.0      17.719000   53.985001        449.0   
9             NaN              0.0      19.164333   60.726332        441.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   
5        456.0               4.0  0.0                      0.0   
6        448.0               4.0  0.0                      0.0   
7        445.0              -4.0  0.0                      0.0   
8        449.0              -8.0  0.0                      0.0   
9        453.0              -8.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  
0                      NaN                      NaN         41.742680  
1                      NaN                      NaN         36.095903  
2                      NaN                      NaN         44.267015  
3                      NaN                      NaN         45.341850  
4                      0.0                      NaN         38.426834  
5                      0.0                      NaN         37.098351  
6                      0.0                      0.0         38.800387  
7                      0.0                      0.0         37.098351  
8                      0.0                      0.0         35.852000  
9                      0.0                      0.0         39.945333

In [14]:
df['soil_moisture_change_rate'] = df['saturation_index'].diff()
print("DataFrame with new 'soil_moisture_change_rate' feature:")
df.head()

DataFrame with new 'soil_moisture_change_rate' feature:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [15]:
print("DataFrame with all engineered features:")
df.head(10)

DataFrame with all engineered features:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700             NaN   
1 2009-01-02            460.0               3.34  0.766881             NaN   
2 2009-01-03            456.0               3.33  0.765062             NaN   
3 2009-01-04            448.0               3.31  0.763244             NaN   
4 2009-01-05            445.0               3.30  0.761425             NaN   
5 2009-01-06            449.0               3.31  0.759606             NaN   
6 2009-01-07            453.0               3.32  0.757787             NaN   
7 2009-01-08            449.0               3.31  0.755969             NaN   
8 2009-01-09            441.0               3.29  0.754150             NaN   
9 2009-01-10            433.0               3.27  0.752331             NaN   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0             NaN              0.0      20.222000   63.263360          NaN   
1             NaN              0.0      18.118000   54.073805        470.0   
2             NaN              0.0      22.784667   65.749364        460.0   
3             NaN              0.0      22.027667   68.656034        456.0   
4             NaN              0.0      18.637667   58.216001        448.0   
5             NaN              0.0      17.945667   56.251036        445.0   
6             NaN              0.0      18.687000   58.913774        449.0   
7             NaN              0.0      17.945667   56.251036        453.0   
8             NaN              0.0      17.719000   53.985001        449.0   
9             NaN              0.0      19.164333   60.726332        441.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   
5        456.0               4.0  0.0                      0.0   
6        448.0               4.0  0.0                      0.0   
7        445.0              -4.0  0.0                      0.0   
8        449.0              -8.0  0.0                      0.0   
9        453.0              -8.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   
5                      0.0                      NaN         37.098351   
6                      0.0                      0.0         38.800387   
7                      0.0                      0.0         37.098351   
8                      0.0                      0.0         35.852000   
9                      0.0                      0.0         39.945333   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016  
5                  -1.328482  
6                   1.702036  
7                  -1.702036  
8                  -1.246351  
9                   4.093333

In [16]:
max_temp_mean = df['Daily Max Temp'].mean()
min_temp_mean = df['Daily Min Temp'].mean()

df['Daily Max Temp'].fillna(max_temp_mean, inplace=True)
df['Daily Min Temp'].fillna(min_temp_mean, inplace=True)

print("DataFrame after imputing NaN values in 'Daily Max Temp' and 'Daily Min Temp':")
df.head()

DataFrame after imputing NaN values in 'Daily Max Temp' and 'Daily Min Temp':


/tmp/ipython-input-2762970626.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Daily Max Temp'].fillna(max_temp_mean, inplace=True)
/tmp/ipython-input-2762970626.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)'

Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          NaN   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [17]:
max_temp_mean = df['Daily Max Temp'].mean()
min_temp_mean = df['Daily Min Temp'].mean()

df['Daily Max Temp'] = df['Daily Max Temp'].fillna(max_temp_mean)
df['Daily Min Temp'] = df['Daily Min Temp'].fillna(min_temp_mean)

print("DataFrame after imputing NaN values in 'Daily Max Temp' and 'Daily Min Temp':")
df.head()

DataFrame after imputing NaN values in 'Daily Max Temp' and 'Daily Min Temp':


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          NaN   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          NaN               NaN  0.0                      NaN   
1          NaN             -10.0  0.0                      NaN   
2          NaN              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [18]:
df['flow_lag_1d'].fillna(0, inplace=True)
df['flow_lag_3d'].fillna(0, inplace=True)

print("DataFrame after imputing NaN values in 'flow_lag_1d' and 'flow_lag_3d' with 0:")
df.head()

DataFrame after imputing NaN values in 'flow_lag_1d' and 'flow_lag_3d' with 0:


/tmp/ipython-input-1177264963.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['flow_lag_1d'].fillna(0, inplace=True)
/tmp/ipython-input-1177264963.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df

Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          0.0   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          0.0               NaN  0.0                      NaN   
1          0.0             -10.0  0.0                      NaN   
2          0.0              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [19]:
df['flow_lag_1d'] = df['flow_lag_1d'].fillna(0)
df['flow_lag_3d'] = df['flow_lag_3d'].fillna(0)

print("DataFrame after imputing NaN values in 'flow_lag_1d' and 'flow_lag_3d' with 0:")
df.head()

DataFrame after imputing NaN values in 'flow_lag_1d' and 'flow_lag_3d' with 0:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          0.0   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          0.0               NaN  0.0                      NaN   
1          0.0             -10.0  0.0                      NaN   
2          0.0              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [20]:
print("DataFrame after all NaN imputations:")
df.head()

DataFrame after all NaN imputations:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          0.0   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          0.0               NaN  0.0                      NaN   
1          0.0             -10.0  0.0                      NaN   
2          0.0              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016

In [21]:
print("DataFrame after all specified NaN imputations:")
df.head(10)

DataFrame after all specified NaN imputations:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   
5 2009-01-06            449.0               3.31  0.759606       26.343369   
6 2009-01-07            453.0               3.32  0.757787       26.343369   
7 2009-01-08            449.0               3.31  0.755969       26.343369   
8 2009-01-09            441.0               3.29  0.754150       26.343369   
9 2009-01-10            433.0               3.27  0.752331       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          0.0   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   
5       16.167989              0.0      17.945667   56.251036        445.0   
6       16.167989              0.0      18.687000   58.913774        449.0   
7       16.167989              0.0      17.945667   56.251036        453.0   
8       16.167989              0.0      17.719000   53.985001        449.0   
9       16.167989              0.0      19.164333   60.726332        441.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          0.0               NaN  0.0                      NaN   
1          0.0             -10.0  0.0                      NaN   
2          0.0              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   
5        456.0               4.0  0.0                      0.0   
6        448.0               4.0  0.0                      0.0   
7        445.0              -4.0  0.0                      0.0   
8        449.0              -8.0  0.0                      0.0   
9        453.0              -8.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      NaN                      NaN         41.742680   
1                      NaN                      NaN         36.095903   
2                      NaN                      NaN         44.267015   
3                      NaN                      NaN         45.341850   
4                      0.0                      NaN         38.426834   
5                      0.0                      NaN         37.098351   
6                      0.0                      0.0         38.800387   
7                      0.0                      0.0         37.098351   
8                      0.0                      0.0         35.852000   
9                      0.0                      0.0         39.945333   

   soil_moisture_change_rate  
0                        NaN  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016  
5                  -1.328482  
6                   1.702036  
7                  -1.702036  
8                  -1.246351  
9                   4.093333

# Task
It looks like the next step is to handle the remaining `NaN` values in the engineered features. Based on the previous steps, several features derived from time-series operations (like `flow_change_rate`, rolling sums, and `soil_moisture_change_rate`) still contain `NaN` values at the beginning of the series.

I will fill these time-series derived `NaN` values with 0, as specified in the plan. I'll also ensure that if `API` or `saturation_index` had any `NaN` values, they would be filled with their respective column means, but based on the current `df.isnull().sum()` output, these columns do not have any `NaN`s.

```python
# Identify columns with remaining NaNs from time-series derivations at the beginning
# Based on df.isnull().sum() from previous step, these are:
# 'flow_change_rate', '24h_accumulated_roll_3d', '24h_accumulated_roll_5d',
# '24h_accumulated_roll_7d', 'soil_moisture_change_rate'

# Fill time-series derived NaNs at the beginning of the series with 0
df['flow_change_rate'] = df['flow_change_rate'].fillna(0)
df['24h_accumulated_roll_3d'] = df['24h_accumulated_roll_3d'].fillna(0)
df['24h_accumulated_roll_5d'] = df['24h_accumulated_roll_5d'].fillna(0)
df['24h_accumulated_roll_7d'] = df['24h_accumulated_roll_7d'].fillna(0)
df['soil_moisture_change_rate'] = df['soil_moisture_change_rate'].fillna(0)

# Verify that all specified NaNs have been handled
print("Remaining NaN values after imputation:")
print(df[['flow_change_rate', '24h_accumulated_roll_3d',
          '24h_accumulated_roll_5d', '24h_accumulated_roll_7d',
          'soil_moisture_change_rate']].isnull().sum())

print("\nFirst 10 rows of the DataFrame after imputing remaining NaNs:")
df.head(10)
```

## Handle Remaining Missing Values

### Subtask:
Impute any remaining NaN values in the engineered features, such as 'flow_change_rate', 'API', rolling sums, and 'soil_moisture_change_rate'. For time-series derived NaNs at the beginning of the series, fill with 0. For other NaNs, fill with the column's mean.


**Reasoning**:
The subtask instructs to fill NaN values in specific columns with 0, as these are likely time-series derived NaNs at the beginning of the series. This includes 'flow_change_rate', '24h_accumulated_roll_3d', '24h_accumulated_roll_5d', '24h_accumulated_roll_7d', and 'soil_moisture_change_rate'.



In [24]:
df['flow_change_rate'] = df['flow_change_rate'].fillna(0)
df['24h_accumulated_roll_3d'] = df['24h_accumulated_roll_3d'].fillna(0)
df['24h_accumulated_roll_5d'] = df['24h_accumulated_roll_5d'].fillna(0)
df['24h_accumulated_roll_7d'] = df['24h_accumulated_roll_7d'].fillna(0)
df['soil_moisture_change_rate'] = df['soil_moisture_change_rate'].fillna(0)

print("Sum of NaN values after imputing with 0:")
print(df[['flow_change_rate', '24h_accumulated_roll_3d', '24h_accumulated_roll_5d', '24h_accumulated_roll_7d', 'soil_moisture_change_rate']].isnull().sum())

print("\nDataFrame after imputing specified NaN values with 0:")
df.head(10)

Sum of NaN values after imputing with 0:
flow_change_rate             0
24h_accumulated_roll_3d      0
24h_accumulated_roll_5d      0
24h_accumulated_roll_7d      0
soil_moisture_change_rate    0
dtype: int64

DataFrame after imputing specified NaN values with 0:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01            470.0               3.36  0.768700       26.343369   
1 2009-01-02            460.0               3.34  0.766881       26.343369   
2 2009-01-03            456.0               3.33  0.765062       26.343369   
3 2009-01-04            448.0               3.31  0.763244       26.343369   
4 2009-01-05            445.0               3.30  0.761425       26.343369   
5 2009-01-06            449.0               3.31  0.759606       26.343369   
6 2009-01-07            453.0               3.32  0.757787       26.343369   
7 2009-01-08            449.0               3.31  0.755969       26.343369   
8 2009-01-09            441.0               3.29  0.754150       26.343369   
9 2009-01-10            433.0               3.27  0.752331       26.343369   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0       16.167989              0.0      20.222000   63.263360          0.0   
1       16.167989              0.0      18.118000   54.073805        470.0   
2       16.167989              0.0      22.784667   65.749364        460.0   
3       16.167989              0.0      22.027667   68.656034        456.0   
4       16.167989              0.0      18.637667   58.216001        448.0   
5       16.167989              0.0      17.945667   56.251036        445.0   
6       16.167989              0.0      18.687000   58.913774        449.0   
7       16.167989              0.0      17.945667   56.251036        453.0   
8       16.167989              0.0      17.719000   53.985001        449.0   
9       16.167989              0.0      19.164333   60.726332        441.0   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0          0.0               0.0  0.0                      0.0   
1          0.0             -10.0  0.0                      0.0   
2          0.0              -4.0  0.0                      0.0   
3        470.0              -8.0  0.0                      0.0   
4        460.0              -3.0  0.0                      0.0   
5        456.0               4.0  0.0                      0.0   
6        448.0               4.0  0.0                      0.0   
7        445.0              -4.0  0.0                      0.0   
8        449.0              -8.0  0.0                      0.0   
9        453.0              -8.0  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      0.0                      0.0         41.742680   
1                      0.0                      0.0         36.095903   
2                      0.0                      0.0         44.267015   
3                      0.0                      0.0         45.341850   
4                      0.0                      0.0         38.426834   
5                      0.0                      0.0         37.098351   
6                      0.0                      0.0         38.800387   
7                      0.0                      0.0         37.098351   
8                      0.0                      0.0         35.852000   
9                      0.0                      0.0         39.945333   

   soil_moisture_change_rate  
0                   0.000000  
1                  -5.646777  
2                   8.171113  
3                   1.074835  
4                  -6.915016  
5                  -1.328482  
6                   1.702036  
7                  -1.702036  
8                  -1.246351  
9                   4.093333

**Reasoning**:
I need to verify if there are any remaining NaN values in the entire DataFrame to ensure all imputation steps of the subtask are completed. If there are no more NaNs, the subtask is finished.



In [25]:
print("Sum of NaN values for all columns:")
print(df.isnull().sum())

Sum of NaN values for all columns:
Date                         0
Daily flow Rate              0
Gauge height flow            0
NDVI                         0
Daily Max Temp               0
Daily Min Temp               0
24h accumulated              0
sm_top_0_10cm                0
sm_10_40cm                   0
flow_lag_1d                  0
flow_lag_3d                  0
flow_change_rate             0
API                          0
24h_accumulated_roll_3d      0
24h_accumulated_roll_5d      0
24h_accumulated_roll_7d      0
saturation_index             0
soil_moisture_change_rate    0
dtype: int64


## Normalize Features

### Subtask:
Scale all relevant numerical features to a common range (e.g., 0 to 1) to ensure that features with larger magnitudes do not disproportionately influence the flood risk score.


**Reasoning**:
First, I'll identify all numerical columns in the DataFrame `df` that need to be scaled, excluding the 'Date' column.



In [26]:
numerical_cols = df.select_dtypes(include=['number']).columns.tolist()
# Exclude 'Date' column if it's in the numerical columns list (though it should be datetime now)
if 'Date' in numerical_cols:
    numerical_cols.remove('Date')

print(f"Numerical columns identified for scaling: {numerical_cols}")

Numerical columns identified for scaling: ['Daily flow Rate', 'Gauge height flow', 'NDVI', 'Daily Max Temp', 'Daily Min Temp', '24h accumulated', 'sm_top_0_10cm', 'sm_10_40cm', 'flow_lag_1d', 'flow_lag_3d', 'flow_change_rate', 'API', '24h_accumulated_roll_3d', '24h_accumulated_roll_5d', '24h_accumulated_roll_7d', 'saturation_index', 'soil_moisture_change_rate']


**Reasoning**:
Now that the numerical columns have been identified, I will import `MinMaxScaler`, initialize it, and then apply it to scale these numerical columns to a range between 0 and 1. The scaled values will replace the original numerical columns in the DataFrame.



In [27]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

print("DataFrame after scaling numerical features using MinMaxScaler:")
df.head()

DataFrame after scaling numerical features using MinMaxScaler:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01         0.015588           0.217881  0.835744        0.511271   
1 2009-01-02         0.015023           0.216379  0.833570        0.511271   
2 2009-01-03         0.014797           0.215627  0.831396        0.511271   
3 2009-01-04         0.014345           0.214125  0.829221        0.511271   
4 2009-01-05         0.014176           0.213373  0.827047        0.511271   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0        0.465784              0.0       0.341041    0.371859     0.000000   
1        0.465784              0.0       0.276175    0.276965     0.026257   
2        0.465784              0.0       0.420048    0.397530     0.025698   
3        0.465784              0.0       0.396709    0.427545     0.025475   
4        0.465784              0.0       0.292196    0.319739     0.025028   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0     0.000000          0.453072  0.0                      0.0   
1     0.000000          0.452654  0.0                      0.0   
2     0.000000          0.452905  0.0                      0.0   
3     0.026257          0.452738  0.0                      0.0   
4     0.025698          0.452947  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      0.0                      0.0          0.362983   
1                      0.0                      0.0          0.275467   
2                      0.0                      0.0          0.402107   
3                      0.0                      0.0          0.418765   
4                      0.0                      0.0          0.311593   

   soil_moisture_change_rate  
0                   0.504389  
1                   0.363425  
2                   0.708370  
3                   0.531221  
4                   0.331765

## Determine Feature Importance (Weights)

### Subtask:
Calculate the absolute Pearson correlation of each normalized feature with the 'Daily flow Rate'. These correlation values will serve as weights, representing each feature's importance in relation to a primary flood indicator.


**Reasoning**:
To calculate the Pearson correlation of each numerical feature with 'Daily flow Rate', I first need to ensure that 'Daily flow Rate' is not correlated with itself and that the 'Date' column is excluded, as specified in the instructions. Then, I will compute the absolute correlations.



In [28]:
numerical_cols_for_correlation = [col for col in numerical_cols if col not in ['Date', 'Daily flow Rate']]

# Calculate Pearson correlation of each feature with 'Daily flow Rate'
correlations = df[numerical_cols_for_correlation].corrwith(df['Daily flow Rate'])

# Take the absolute value of the correlation coefficients
feature_weights = correlations.abs()

print("Feature Importance Weights (Absolute Pearson Correlation with 'Daily flow Rate'):")
print(feature_weights)


Feature Importance Weights (Absolute Pearson Correlation with 'Daily flow Rate'):
Gauge height flow            0.963999
NDVI                         0.058615
Daily Max Temp               0.191910
Daily Min Temp               0.164385
24h accumulated              0.069141
sm_top_0_10cm                0.662400
sm_10_40cm                   0.675176
flow_lag_1d                  0.932997
flow_lag_3d                  0.847708
flow_change_rate             0.183051
API                          0.116020
24h_accumulated_roll_3d      0.094641
24h_accumulated_roll_5d      0.109113
24h_accumulated_roll_7d      0.114186
saturation_index             0.675935
soil_moisture_change_rate    0.012638
dtype: float64


## Calculate Weighted Flood Risk Score

### Subtask:
Compute a composite 'Weighted Flood Risk Score' for each day by summing the product of each normalized feature and its corresponding importance weight.


**Reasoning**:
As instructed, I will initialize a new column 'Weighted Flood Risk Score' in the DataFrame `df` with zeros. This column will accumulate the weighted sum of each normalized feature.



In [29]:
df['Weighted Flood Risk Score'] = 0.0
print("DataFrame with initialized 'Weighted Flood Risk Score' column:")
df.head()

DataFrame with initialized 'Weighted Flood Risk Score' column:


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01         0.015588           0.217881  0.835744        0.511271   
1 2009-01-02         0.015023           0.216379  0.833570        0.511271   
2 2009-01-03         0.014797           0.215627  0.831396        0.511271   
3 2009-01-04         0.014345           0.214125  0.829221        0.511271   
4 2009-01-05         0.014176           0.213373  0.827047        0.511271   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0        0.465784              0.0       0.341041    0.371859     0.000000   
1        0.465784              0.0       0.276175    0.276965     0.026257   
2        0.465784              0.0       0.420048    0.397530     0.025698   
3        0.465784              0.0       0.396709    0.427545     0.025475   
4        0.465784              0.0       0.292196    0.319739     0.025028   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0     0.000000          0.453072  0.0                      0.0   
1     0.000000          0.452654  0.0                      0.0   
2     0.000000          0.452905  0.0                      0.0   
3     0.026257          0.452738  0.0                      0.0   
4     0.025698          0.452947  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      0.0                      0.0          0.362983   
1                      0.0                      0.0          0.275467   
2                      0.0                      0.0          0.402107   
3                      0.0                      0.0          0.418765   
4                      0.0                      0.0          0.311593   

   soil_moisture_change_rate  Weighted Flood Risk Score  
0                   0.504389                        0.0  
1                   0.363425                        0.0  
2                   0.708370                        0.0  
3                   0.531221                        0.0  
4                   0.331765                        0.0

**Reasoning**:
Now that the 'Weighted Flood Risk Score' column is initialized, I will iterate through the numerical features (excluding 'Daily flow Rate'), multiply each feature's normalized values by its corresponding absolute Pearson correlation weight, and sum these products into the 'Weighted Flood Risk Score' column. Finally, I will display the first few rows to verify the calculation.



In [30]:
for feature in numerical_cols_for_correlation:
    df['Weighted Flood Risk Score'] += df[feature] * feature_weights[feature]

print("DataFrame after calculating 'Weighted Flood Risk Score':")
df.head()

DataFrame after calculating 'Weighted Flood Risk Score':


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01         0.015588           0.217881  0.835744        0.511271   
1 2009-01-02         0.015023           0.216379  0.833570        0.511271   
2 2009-01-03         0.014797           0.215627  0.831396        0.511271   
3 2009-01-04         0.014345           0.214125  0.829221        0.511271   
4 2009-01-05         0.014176           0.213373  0.827047        0.511271   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0        0.465784              0.0       0.341041    0.371859     0.000000   
1        0.465784              0.0       0.276175    0.276965     0.026257   
2        0.465784              0.0       0.420048    0.397530     0.025698   
3        0.465784              0.0       0.396709    0.427545     0.025475   
4        0.465784              0.0       0.292196    0.319739     0.025028   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0     0.000000          0.453072  0.0                      0.0   
1     0.000000          0.452654  0.0                      0.0   
2     0.000000          0.452905  0.0                      0.0   
3     0.026257          0.452738  0.0                      0.0   
4     0.025698          0.452947  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      0.0                      0.0          0.362983   
1                      0.0                      0.0          0.275467   
2                      0.0                      0.0          0.402107   
3                      0.0                      0.0          0.418765   
4                      0.0                      0.0          0.311593   

   soil_moisture_change_rate  Weighted Flood Risk Score  
0                   0.504389                   1.245349  
1                   0.363425                   1.100220  
2                   0.708370                   1.365557  
3                   0.531221                   1.399827  
4                   0.331765                   1.181143

## Assign Flood Risk Labels

### Subtask:
Discretize the 'Weighted Flood Risk Score' into four distinct classes (0, 1, 2, 3) to represent different levels of flood risk. This can be done using quantiles to create balanced classes or by defining specific thresholds.


**Reasoning**:
First, I will define the number of classes for the flood risk score and then use `pd.qcut()` to discretize the 'Weighted Flood Risk Score' into these classes, assigning the resulting labels to a new column 'Flood Risk Label'.



In [31]:
num_classes = 4
df['Flood Risk Label'] = pd.qcut(df['Weighted Flood Risk Score'], q=num_classes, labels=False, duplicates='drop')

print(f"Value counts for 'Flood Risk Label' with {num_classes} classes:")
print(df['Flood Risk Label'].value_counts().sort_index())

print("\nFirst 5 rows of the DataFrame with 'Flood Risk Label':")
df.head()

Value counts for 'Flood Risk Label' with 4 classes:
Flood Risk Label
0    1370
1    1369
2    1369
3    1370
Name: count, dtype: int64

First 5 rows of the DataFrame with 'Flood Risk Label':


Date  Daily flow Rate  Gauge height flow      NDVI  Daily Max Temp  \
0 2009-01-01         0.015588           0.217881  0.835744        0.511271   
1 2009-01-02         0.015023           0.216379  0.833570        0.511271   
2 2009-01-03         0.014797           0.215627  0.831396        0.511271   
3 2009-01-04         0.014345           0.214125  0.829221        0.511271   
4 2009-01-05         0.014176           0.213373  0.827047        0.511271   

   Daily Min Temp  24h accumulated  sm_top_0_10cm  sm_10_40cm  flow_lag_1d  \
0        0.465784              0.0       0.341041    0.371859     0.000000   
1        0.465784              0.0       0.276175    0.276965     0.026257   
2        0.465784              0.0       0.420048    0.397530     0.025698   
3        0.465784              0.0       0.396709    0.427545     0.025475   
4        0.465784              0.0       0.292196    0.319739     0.025028   

   flow_lag_3d  flow_change_rate  API  24h_accumulated_roll_3d  \
0     0.000000          0.453072  0.0                      0.0   
1     0.000000          0.452654  0.0                      0.0   
2     0.000000          0.452905  0.0                      0.0   
3     0.026257          0.452738  0.0                      0.0   
4     0.025698          0.452947  0.0                      0.0   

   24h_accumulated_roll_5d  24h_accumulated_roll_7d  saturation_index  \
0                      0.0                      0.0          0.362983   
1                      0.0                      0.0          0.275467   
2                      0.0                      0.0          0.402107   
3                      0.0                      0.0          0.418765   
4                      0.0                      0.0          0.311593   

   soil_moisture_change_rate  Weighted Flood Risk Score  Flood Risk Label  
0                   0.504389                   1.245349                 1  
1                   0.363425                   1.100220                 1  
2                   0.708370                   1.365557                 1  
3                   0.531221                   1.399827                 1  
4                   0.331765                   1.181143                 1

## Review Flood Risk Labels

### Subtask:
Display the distribution of the newly created 'Flood Risk Label' column and the first few rows of the DataFrame, including the new score and labels, to verify the outcome.


**Reasoning**:
As instructed, I will display the value counts of the 'Flood Risk Label' column to show its distribution and then the first few rows of the DataFrame, focusing on 'Weighted Flood Risk Score' and 'Flood Risk Label' columns.



In [32]:
print("Distribution of 'Flood Risk Label':")
print(df['Flood Risk Label'].value_counts().sort_index())

print("\nFirst few rows of DataFrame with 'Weighted Flood Risk Score' and 'Flood Risk Label':")
print(df[['Weighted Flood Risk Score', 'Flood Risk Label']].head())

Distribution of 'Flood Risk Label':
Flood Risk Label
0    1370
1    1369
2    1369
3    1370
Name: count, dtype: int64

First few rows of DataFrame with 'Weighted Flood Risk Score' and 'Flood Risk Label':
   Weighted Flood Risk Score  Flood Risk Label
0                   1.245349                 1
1                   1.100220                 1
2                   1.365557                 1
3                   1.399827                 1
4                   1.181143                 1


## Final Task

### Subtask:
Summarize the process of creating flood risk labels, including the chosen method for feature importance and weighting, and provide insights into the resulting classification.


## Summary:

### Q&A
*   **What was the chosen method for feature importance and weighting?**
    The chosen method for feature importance and weighting was the absolute Pearson correlation of each normalized feature with the 'Daily flow Rate'. These correlation values were then used as weights.

*   **What insights can be drawn into the resulting classification?**
    The resulting classification produced four distinct flood risk labels (0, 1, 2, 3) which are well-balanced across the dataset due to the use of quantile-based discretization. This indicates a robust distribution of risk levels, with each class representing approximately 25% of the data.

### Data Analysis Key Findings
*   All remaining missing values in engineered features were successfully handled: `NaN` values in time-series derived features (`flow_change_rate`, `24h_accumulated_roll_3d`, `24h_accumulated_roll_5d`, `24h_accumulated_roll_7d`, `soil_moisture_change_rate`) were filled with 0. No `NaN`s were found in `API` or `saturation_index` requiring mean imputation.
*   17 numerical features, including `Daily flow Rate`, `Gauge height flow`, and various accumulated and change rate features, were successfully scaled to a range between 0 and 1 using `MinMaxScaler`.
*   Feature importance weights were calculated using the absolute Pearson correlation with 'Daily flow Rate'. Key features such as 'Gauge height flow' (0.964), 'flow_lag_1d' (0.933), 'flow_lag_3d' (0.848), 'sm_10_40cm' (0.675), and 'saturation_index' (0.676) showed strong correlations, indicating their high importance in predicting flood risk. Features like 'NDVI' (0.059) and 'soil_moisture_change_rate' (0.013) showed very low correlation.
*   A 'Weighted Flood Risk Score' was computed for each day by summing the product of each normalized feature and its corresponding importance weight.
*   The 'Weighted Flood Risk Score' was discretized into four 'Flood Risk Label' classes (0, 1, 2, 3) using quantile-based categorization (`pd.qcut`). The distribution of these labels was highly balanced, with 1370 entries for labels 0 and 3, and 1369 entries for labels 1 and 2.

### Insights or Next Steps
*   The use of absolute Pearson correlation with 'Daily flow Rate' as a weighting mechanism effectively highlights features most linearly associated with flood events, providing a transparent and interpretable basis for the flood risk score.
*   Future work could involve exploring non-linear methods for feature importance (e.g., tree-based model feature importance) or dynamic weighting schemes to capture more complex relationships and potentially improve flood risk prediction accuracy, especially for extreme events.
